core/tracestack.lua

```diff
@@ -1,147 +1,137 @@
+  -- Push a command to the stack to record the execution trace for debugging.
   -- Carries information about the command call, not the command itself.
   -- Must be popped with `pop(returnOfPush)`.
-  pushCommand = function(self, tag, line, column, options, file)
+  pushCommand = function(self, tag, options, file, line, column)
     if not tag then
       SU.warn("Tag should be specified for SILE.traceStack:pushCommand", true)
     end
-    return self:_pushFrame({
-        tag = tag or "???",
-        file = file or SILE.currentlyProcessingFile,
-        line = line,
-        column = column,
-        options = options or {}
-      })
+    local file = file or SILE.currentlyProcessingFile
+    return self:_push(file, line, column, tag, options)
   end,
 
 
-  -- Push a command frame on to the stack to record the execution trace for debugging.
+  -- Push a command to the stack to record the execution trace for debugging.
   -- Command arguments are inferred from AST content, any item may be overridden.
   -- Must be popped with `pop(returnOfPush)`.
-  pushContent = function(self, content, tag, line, column, options, file)
+  pushContent = function(self, content, tag)
+    local tag = tag or content.tag
     if type(content) ~= "table" then
       SU.warn("Content parameter of SILE.traceStack:pushContent must be a table", true)
       content = {}
     end
-    tag = tag or content.tag
     if not tag then
       SU.warn("Tag should be specified or inferable for SILE.traceStack:pushContent", true)
     end
-    return self:_pushFrame({
-        tag = tag or "???",
-        file = file or content.file or SILE.currentlyProcessingFile,
-        line = line or content.line,
-        column = column or content.col,
-        options = options or content.attr or {}
-      })
+    local file = content.file or SILE.currentlyProcessingFile
+    return self:_push(file, content.line, content.col, tag, content.attr)
   end,
 
   -- Push a text that is going to get typeset on to the stack to record the execution trace for debugging.
   -- Must be popped with `pop(returnOfPush)`.
-  pushText = function(self, text, line, column, file)
-    return self:_pushFrame({
-        text = text,
-        file = file,
-        line = line,
-        column = column
-      })
+  pushText = function(self, text)
+    return self:_push(nil, nil, nil, nil, nil, text)
   end,
 
-  -- Internal: Push complete frame
-  _pushFrame = function (self, frame)
-    frame.toLocationString = self._frameToLocationString
-    frame.typesetter = SILE.typesetter
-    -- Push the frame
-    if SU.debugging("commandStack") then
-      print(string.rep(" ", #self) .. "PUSH(" .. frame:toLocationString(false, true) .. ")")
-    end
-    self[#self + 1] = frame
-    self.afterFrame = nil
-    self._lastPushId = self._lastPushId + 1
-    frame.pushId = self._lastPushId
-    return self._lastPushId
+  _push = function (self, file, line, column, tag, options, text)
+    local pushId = #self + 1
```

Comment:
This will not catch as many errors as the code that was here previously.

Reply:
Why? All I'm doing is normalizing the ID using Lua's internal table key value so we don't end up with things in the table at a different location than our representation of them claims they should be. You were just iterating the number yourself, but storing it in an arbitrary and possibly not matching table location. What error would you catch by doing that? What is a test case I could run that would make my id system not work correctly?

Reply:
You use the stack height as the push ID (which will repeat), but I have used monotonically increasing ID. Consider this sequence of commands:
```
PUSH(a) => 1
PUSH(b) => 2
POP(2) = b

PUSH(c) => 2

-- PUSH(b) -- But this call is for some reason forgotten, but the code still assumes that it was made
POP(2) -- By the code which forgot to call PUSH previously, but still remembers pushID = 2

POP(2) => Fail -- By the code which pushed c
```

The previous system would catch this earlier, as it never reuses IDs. I agree that this is an convoluted example, but there could be a better one I can't think of.

Reply:
Okay I see what you are saying, but if you really want to catch that case then, ironically, _both_ of our implementations turn out to be buggy! For different reasons, both of us were allowing the table keys to be non-contiguous. That means we cannot naively use `for i = ..., #tbl` to iterate over the table.

Lua would actually have covered my tail in the particular example you give because the `#tbl` operator does not always give the depth of the stack but instead the highest contiguous key it finds, so it would not allow me to reuse keys any more than you were. But that would still allow unexpected results when iterating over it, so I've gone through and fixed that.

Reply:
What is the bug in my implementation? I haven't used the `pushId` as a key to the stack. And I can't think of a way to force non-contiguity in my implementation, in any way, as it literally does just `self[#self + 1] = frame` and `self[#self] = nil`.

Either I am missing something obvious, or you might have misread how the original implementation worked.